tsl/src/bgw_policy/scheduled_index_api.c

```diff
@@ -0,0 +1,245 @@
+	if (!ts_is_hypertable(ht_oid))
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_HYPERTABLE_NOT_EXIST),
+				 errmsg("could not add scheduled_index policy because \"%s\" is not a hypertable",
+						get_rel_name(ht_oid))));
+
+	/* Now verify that the index is an actual index on that hypertable */
+	check_valid_index(ht, index_name);
+
+	/* Make sure that an existing policy doesn't exist on this hypertable */
+	existing = ts_bgw_policy_scheduled_index_find_by_hypertable(ts_hypertable_relid_to_id(ht_oid));
+
+	if (existing != NULL)
+	{
+		if (!if_not_exists)
+			ereport(ERROR,
+					(errcode(ERRCODE_DUPLICATE_OBJECT),
+					 errmsg("scheduled_index policy already exists for hypertable \"%s\"",
+							get_rel_name(ht_oid))));
+
+		if (!DatumGetBool(DirectFunctionCall2(nameeq,
+											  NameGetDatum(&existing->fd.hypertable_index_name),
+											  NameGetDatum(index_name))))
+		{
+			elog(WARNING,
+				 "could not add scheduled_index policy due to existing policy on hypertable with "
+				 "different "
+				 "arguments");
+			return -1;
+		}
+		/* If all arguments are the same, do nothing */
+		ereport(NOTICE,
+				(errmsg("scheduled_index policy already exists on hypertable \"%s\", skipping",
+						get_rel_name(ht_oid))));
+		return -1;
+	}
+
+	/* Next, insert a new job into jobs table */
+	namestrcpy(&application_name, "ScheduledIndex Background Job");
+	namestrcpy(&scheduled_index_name, "scheduled_index");
+
+	/*
+	 * Try to see if the hypertable has a specified chunk length for the
+	 * default schedule interval
+	 */
+	dim = hyperspace_get_open_dimension(ht->space, 0);
+
+	if (dim && IS_TIMESTAMP_TYPE(dim->fd.column_type))
+		default_schedule_interval = DatumGetIntervalP(
+			DirectFunctionCall7(make_interval,
+								Int32GetDatum(0),
+								Int32GetDatum(0),
+								Int32GetDatum(0),
+								Int32GetDatum(0),
+								Int32GetDatum(0),
+								Int32GetDatum(0),
+								Float8GetDatum(dim->fd.interval_length / 2000000)));
+
+	job_id = ts_bgw_job_insert_relation(&application_name,
+										&scheduled_index_name,
+										default_schedule_interval,
+										DEFAULT_MAX_RUNTIME,
+										DEFAULT_MAX_RETRIES,
+										DEFAULT_RETRY_PERIOD);
+
+	/* Now, insert a new row in the scheduled_index args table */
+	policy.fd.job_id = job_id;
+	ts_bgw_policy_scheduled_index_insert(&policy);
+
+	optional_index_info = ts_indexing_optional_info_find_by_index_name(index_name);
+	if (optional_index_info == NULL)
+		optional_index_info = ts_optional_index_info_alloc(NameStr(*index_name));
+
+	optional_index_info->fd.is_scheduled = true;
+	Assert(DirectFunctionCall2(nameeq,
+							   NameGetDatum(index_name),
+							   NameGetDatum(&optional_index_info->fd.hypertable_index_name)));
+
+	// TODO: use update when optional_index_info already exists
+	ts_indexing_optional_info_insert(optional_index_info);
+
+	PG_RETURN_INT32(job_id);
+}
+
+Datum
+scheduled_index_remove_policy(PG_FUNCTION_ARGS)
```

Comment:
if a policy is removed shouldn't we add indexes to any chunks that don't have them?

Reply:
This should take a particular hypertable_index